### PR TITLE
Refactor `MixedBulkWriteOperation` a bit and change `CommandMessage.isResponseExpected` such that it accounts for ordered/unordered bulk writes

### PR DESCRIPTION
**Pull request by stIncMale**

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java [hunk 1]
```diff
@@ -101,12 +101,7 @@ static BulkWriteBatch createBulkWriteBatch(final MongoNamespace namespace,
                                                       final List<? extends WriteRequest> writeRequests,
                                                       final OperationContext operationContext,
                                                       @Nullable final BsonValue comment, @Nullable final BsonDocument variables) {
-        SessionContext sessionContext = operationContext.getSessionContext();
-        if (sessionContext.hasSession() && !sessionContext.isImplicitSession() && !sessionContext.hasActiveTransaction()
-                && !writeConcern.isAcknowledged()) {
-            throw new MongoClientException("Unacknowledged writes are not supported when using an explicit session");
-        }
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java:108]: This validation was moved to `MixedBulkWriteOperation.validateAndGetEffectiveWriteConcern`.

File: driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java [hunk 2]
```diff
@@ -169,9 +164,8 @@ private BulkWriteBatch(final MongoNamespace namespace, final ConnectionDescripti
         if (!payloadItems.isEmpty()) {
             command.put(getCommandName(batchType), new BsonString(namespace.getCollectionName()));
             command.put("ordered", new BsonBoolean(ordered));
-            if (!writeConcern.isServerDefault() && !sessionContext.hasActiveTransaction()) {
-                command.put("writeConcern", writeConcern.asDocument());
-            }
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java:174]: This logic needs to be reused in the implementation of the new `bulkWrite` command, so I extracted it into `MixedBulkWriteOperation.commandWriteConcern`.

File: driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java [hunk 3]
```diff
@@ -405,31 +414,47 @@ private boolean handleMongoWriteConcernWithResponseExceptionAsync(final RetrySta
...
-                : getAppliedWriteConcern(sessionContext).isAcknowledged();
+    static Optional<WriteConcern> commandWriteConcern(final WriteConcern effectiveWriteConcern, final SessionContext sessionContext) {
+        return effectiveWriteConcern.isServerDefault() || sessionContext.hasActiveTransaction()
+                ? Optional.empty()
+                : Optional.of(effectiveWriteConcern);
+    }
+
+    private boolean shouldExpectResponse(final BulkWriteBatch batch, final WriteConcern effectiveWriteConcern) {
+        return effectiveWriteConcern.isAcknowledged() || (ordered && batch.hasAnotherBatch());
     }
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java:458]: `shouldExpectResponse` replaces the `shouldAcknowledge` method. The new method has simpler body and named in a way that avoids erroneous allusions to an unacknowledged (`w: 0`) write concern.

File: driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java [hunk 4]
```diff
@@ -253,7 +253,7 @@ private boolean requireOpMsgResponse() {
         if (responseExpected) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java:253]: We can simplify this class bit further by adding the following assertion in the constructor:

```
assertFalse(getOpCode() == OpCode.OP_QUERY && !responseExpected);
```

this assertion should never fire because OP_QUERY effectively is only used for the initial `hello` command on a new connection, which should always be expecting a response.

Then we can remove the `!useOpMsg()` condition in `isResponseExpected`, and inline this method.

File: driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java (hunk 2 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java:174]: Will it eventually move to some common "helper" class?

Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java:174]: I had no such plans, as I don't see any value in creating a somewhat useless class just to hold this method, such that both `MixedBulkWriteOperation` and `ClientBulkWriteOperation` could call `commandWriteConcern` on that class. Calling `commandWriteConcern` on `MixedBulkWriteOperation` seems to be not worse in any way, especially given that when it is statically imported, one can't even notice immediately see where the method comes from.

File: driver-sync/src/test/functional/com/mongodb/client/MongoClientSessionSpecification.groovy [hunk 5]
```diff
@@ -350,9 +351,11 @@ class MongoClientSessionSpecification extends FunctionalSpecification {
 
     void waitForInsertAcknowledgement(MongoCollection<Document> collection, ObjectId id) {
         Document document = collection.find(Filters.eq(id)).first()
+        Timeout timeout = Timeout.expiresIn(5, TimeUnit.SECONDS, Timeout.ZeroSemantics.ZERO_DURATION_MEANS_INFINITE)
```
Comment (jyemin) [inline driver-sync/src/test/functional/com/mongodb/client/MongoClientSessionSpecification.groovy:354]: Is this change relevant to the rest of the PR?

**Review by jyemin — APPROVED** [review]

File: driver-sync/src/test/functional/com/mongodb/client/MongoClientSessionSpecification.groovy (hunk 5 above)
Comment (stIncMale) [reply driver-sync/src/test/functional/com/mongodb/client/MongoClientSessionSpecification.groovy:354]: Sorry, I thought I posted my comments, but they were pending. Yes, here are the relevant comments:
- https://github.com/mongodb/mongo-java-driver/pull/1481#discussion_r1718712117
- https://github.com/mongodb/mongo-java-driver/pull/1481#discussion_r1718764877
